File: 백준/Silver/15650. N과 M （2）/N과 M （2）.py
from sys import stdin
input = stdin.readline

def solution(start: int, depth: int):
    if depth == M:
        print(*arr)
        return

    for i in range(start, N):
        if not visit[i]:
            visit[i] = True
            arr[depth] = i + 1
            solution(i + 1, depth + 1)
            visit[i] = False

N, M = map(int, input().split())
arr = [0] * M
visit = [False] * N
solution(0, 0)

File: 백준/Silver/15650. N과 M （2）/N과 M （2）.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Main {
    static int N, M;
    static int[] arr;
    static boolean[] visit;
    static StringBuilder sb = new StringBuilder();

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        br.close();

        N = Integer.parseInt(st.nextToken());
        M = Integer.parseInt(st.nextToken());

        arr = new int[M];
        visit = new boolean[N];

        solution(0, 0);

        System.out.println(sb);
    }

    static void solution(int start, int depth) {
        if (depth == M) {
            for (int i : arr) {
                sb.append(i).append(' ');
            }
            sb.append('\n');
            return;
        }

        for (int i = start; i < N; i++) {
            if (!visit[i]) {
                visit[i] = true;
                arr[depth] = i + 1;
                solution(i + 1, depth + 1);
                visit[i] = false;
            }
        }
    }
}